### ISSUE TITLE
Hot updates not applied

### ISSUE THREAD

**Issue by egorovli** [gaearon/react-hot-loader#581]
I'm currently experiencing a weird behavior, hot updates stopped working in my project some commits ago, and I can't figure out what caused it. Currently I'll ask a quick question in case anyone knows, if not, I'll try to reproduce it in an isolated environment.

Ok, everything seemed to work, but now hot updates stopped applying, however, console says that it's working:

<img width="610" alt="screen shot 2017-06-10 at 00 13 51" src="https://user-images.githubusercontent.com/5165202/26994772-bdb74200-4d71-11e7-8d0e-480459073916.png">

<img width="860" alt="screen shot 2017-06-10 at 00 18 13" src="https://user-images.githubusercontent.com/5165202/26994866-5a6c30d8-4d72-11e7-8aa7-87e0367b8fdd.png">

I'm using `"react-hot-loader": "^3.0.0-beta.6"`, serving files from `"webpack-dev-server": "^2.4.5"`:

```javascript
// ...
entry: {
  admin: [
    'babel-polyfill',
    ...(env.dev ? [
      'react-hot-loader/patch'
    ] : []),
    './admin/polyfills.js',
    './styles/admin.scss',
    './admin/index.js'
  ],
  client: [
    'babel-polyfill',
    ...(env.dev ? [
      'react-hot-loader/patch'
    ] : []),
    './client/polyfills.js',
    './styles/client.scss',
    './client/index.js'
  ]
},
output: {
  path: distPath,
  filename: env.prod ?
    'js/[name].[chunkhash:5].min.js' :
    'js/[name].js',
  publicPath: 'http://localhost:8081/dist/'
},
devServer: {
  hot: true,
  compress: true,
  overlay: true,
  port: 8081,
  headers: {
    'Access-Control-Allow-Origin': '*'
  }
}
// ...
```

So basically all files are there at `http://localhost:8081/dist/js/admin.js` and `client.js` (entry points), all hot updates are there, as well, client app knows it should update and what updates should be applied (see screenshots of the console above), but nothing changes in the browser.

Any help is very much appreciated.

**Comment by egorovli** [comment]
Ok, I've created a simplified repo of my project: https://github.com/egorovli/react-hmr-test
Video of hot updates not applied: https://www.dropbox.com/s/uuav7tpz9d1l35s/HMR%20Not%20Working.mov?dl=0

**Comment by egorovli** [comment]
This one is related: https://github.com/webpack/webpack-dev-server/issues/100

I managed to fix the issue by changing

```javascript
if (module.hot) {
  module.hot.accept('./containers/Root', () => render(Root));
}
```

to

```javascript
if (module.hot) {
  module.hot.accept('./containers/Root', () => render(require('./containers/Root').default));
}
```

Not sure if it's an intended behavior. https://github.com/gaearon/react-hot-boilerplate and https://github.com/wkwiatek/react-hot-loader-minimal-boilerplate both use this syntax, as well as official docs https://github.com/gaearon/react-hot-loader/tree/master/docs:

```javascript
if (module.hot) module.hot.accept('./App', () => render(App));
```

**Comment by egorovli** [comment]
Or you can use:

```javascript
if (module.hot) {
  module.hot.accept();
}
```

But in my case this was throwing errors about `store`. In non-redux environment, like mobx, it works fine.

**Comment by QuietWind** [comment]
`  javascript
	const reactAppContainer = document.getElementById('reactAppContainer')
	if (module.hot) {
		module.hot.accept()
		ReactDOM.render((<AppContainer><MainPage /></AppContainer>), reactAppContainer)
	} else {
		ReactDOM.render((<AppContainer><MainPage /></AppContainer>), reactAppContainer)
	}
`

appContainer use this. and setting with [migration-to-3.0](https://github.com/gaearon/react-hot-loader/tree/master/docs#migration-to-30)

**Comment by chrisregner** [comment]
Another case that may cause this is *not* having the `react-hot-loader/babel` babel plugin as the first element of array in the config file. @gaearon, is this behavior normal? If so I'd like to make a PR on documentation so that fewer people get tripped over this (I kinda spent plenty of time just because of this).

**Comment by tvedtorama** [comment]
@egorovli I spent hours to trying to get my rhl 3 update to correctly apply edits, then your solution with *module.hot.accept()* worked. The recommended way, with  *module.hot.accept(path, callback)* failed silently - with "Component does not know how to update itself".  I think there must be something about the relative paths or anything in my typescript setup that fails it. 

Is there any documentation/links on what this does, exactly - since we don't get the chance to explicitly call the render function again?  Apart from the very annoying store error, are there any other disadvantages to using this simple syntax?

**Comment by vladimirdh** [comment]
@egorovli this also works for me, with react v16 and webpack v3

**Comment by neoziro** [comment]
Clarified in https://github.com/gaearon/react-hot-loader#getting-started

**Comment by yuqingc** [comment]
@egorovli Thanks so much for your solution. I solved this problem with the help of your _require_ and _default_ solution. Do you know how that happened? Is there some error about the babel config?

**Comment by yofi2tofi** [comment]
@egorovli thank you so muck. I don't know why, but i had wrong behavior when set code like of. doc

**Comment by dangdennis** [comment]
Wow @egorovli 

I entered a new project recently, and noticed that our HMR wasn't working properly. No idea why, thought it might've been the way browsersync was set up.

Regardless, simply using

```
 if (module.hot) {
  module.hot.accept();
}
```
did the trick. Still haven't figured out why though.

**Comment by tommyalvarez** [comment]
@egorovli your solution also worked for me. Using React 16.3 and webpack v4.  This is my main index.js entry point file:

`import React from 'react'
import ReactDOM from 'react-dom'
import createStore from './store/createStore'
import { hot } from 'react-hot-loader'
import AppContainer from './containers/AppContainer'
import { persistStore } from 'redux-persist'

// ========================================================
// Store Instantiation
// ========================================================

const initialState = window.___INITIAL_STATE__
const { history, store } = createStore(initialState)

// begin periodically persisting the store
let persistor = persistStore(store)

// ========================================================
// Render Setup
// ========================================================
const MOUNT_NODE = document.getElementById('root')

let render = () => {
  ReactDOM.render(
    <AppContainer store={store} persistor={persistor} history={history} />,
    MOUNT_NODE
  )
}

// This code is excluded from production bundle
if (__DEV__) {
  if (module.hot) {
    // Development render functions
    const renderApp = render
    const renderError = (error) => {
      const RedBox = require('redbox-react').default

      ReactDOM.render(<RedBox error={error} />, MOUNT_NODE)
    }

    // Wrap render in try/catch
    render = () => {
      try {
        renderApp()
      } catch (error) {
        console.error(error)
        renderError(error)
      }
    }

    // Setup hot module replacement
    module.hot.accept('./containers/AppContainer', () =>
      render(require('./containers/AppContainer').default)
    )
  }
}

// ========================================================
// Go!
// ========================================================
export default hot(module)(render)
render()
`

**Comment by theKashey** [comment]
@dangdennis - that might be due to a bug in React-Hot-Loader v4, when it does nothing without babel (#923)

**Comment by WangYang-Rex** [comment]
how to use 'AppContainer'  with 'react-router' and 'redux‘ together?

**Comment by theKashey** [comment]
Always use just `hot`.

**Comment by WangYang-Rex** [comment]
can you provide some example?

**Comment by theKashey** [comment]
Any our example - just wrap with `hot` some component "just before" react-dom render, and it should be an export of another file
